src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -627,8 +627,8 @@ else if (command.getRenewToken() != null) {
 							RelayLogger.logEvent("sendCommand", this, json);
 							return webSocket.writeAsync(json, timeout, WriteMode.TEXT)
 								.thenCompose($void -> {
-									if (buffer != null) {
-										return webSocket.writeAsync(buffer.array());
+									if (writeBufferStream != null) {
+										return webSocket.writeAsync(writeBufferStream.toByteArray());
```

Comment:
>toByteArray [](start = 56, length = 11)

Ugh, I don't think this is any different than how it was before but any time we go from a ByteBuffer or Stream to an array of the exact right size we're basically making it impossible to pool those buffers.

Reply:
There is nothing different here, just changing the signature of the function for convenience sake.

**writeAsync()** only takes byte[] or ByteBuffer, so I don't think there is a better way. (I recall trying their OutputStream way but it just did not work, can try again)

---
In reply to: [276023248](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276023248) [](ancestors = 276023248)

Reply:
For future consideration one could subclass ByteArrayOutputStream in order to provide access to the underlying byte[] without copying into a new one, and then create a ByteBuffer using that array and the size() for passing into webSocket.writeAsync. This would avoid allocating a new array and having to copy all the data.

---
In reply to: [276041048](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276041048) [](ancestors = 276041048,276023248)

Reply:
It can just use ByteBuffer like before, just need to manually allocate a new, larger ByteBuffer when needed (ByteArrayOutputStream needs to resize anyways).

---
In reply to: [276301369](https://github.com/Azure/azure-relay-java/pull/39#discussion_r276301369) [](ancestors = 276301369,276041048,276023248)